tests/framework/icd/test_icd.h

```diff
@@ -0,0 +1,158 @@
+ *
+ * Author: Charles Giessen <charles@lunarg.com>
+ */
+
+#pragma once
+
+#include "util/common.h"
+#include "util/test_util.h"
+
+#include "layer/layer_common.h"
+
+#include "physical_device.h"
+
+
+enum class CalledICDGIPA { not_called, vk_icd_gipa, vk_gipa };
+
+enum class CalledNegotiateInterface { not_called, vk_icd_negotiate, vk_icd_gipa_first };
+
+enum class InterfaceVersionCheck { not_called, loader_version_too_old, loader_version_too_new, icd_version_too_new, version_is_supported};
+
+enum class CalledEnumerateAdapterPhysicalDevices { not_called, called, called_but_not_supported};
+
+enum class UsingICDProvidedWSI { not_using, is_using};
+
+struct TestICD {
+    fs::path manifest_file_path;
+
+    CalledICDGIPA called_vk_icd_gipa = CalledICDGIPA::not_called;
+    CalledNegotiateInterface called_negotiate_interface = CalledNegotiateInterface::not_called;
+
+    InterfaceVersionCheck interface_version_check = InterfaceVersionCheck::not_called;
+    uint32_t min_icd_interface_version = 0; 
+    uint32_t max_icd_interface_version = 6;
+    uint32_t icd_interface_version_received = 0;
+
+    CalledEnumerateAdapterPhysicalDevices called_enumerate_adapter_physical_devices = CalledEnumerateAdapterPhysicalDevices::not_called;
+
+    bool enable_icd_wsi = false;
+    UsingICDProvidedWSI is_using_icd_wsi = UsingICDProvidedWSI::not_using;
+
+    uint32_t icd_api_version = VK_MAKE_VERSION(1, 0, 0);
+    std::vector<Layer> instance_layers;
+    std::vector<Extension> instance_extensions;
+    std::vector<PhysicalDevice> physical_devices;
+
+    DispatchableHandle<VkInstance> instance_handle;
+    std::vector<DispatchableHandle<VkDevice>> device_handles;
+
+    uint64_t created_surface_count = 0;
+    uint64_t created_swapchain_count = 0;
+
+
+    TestICD() {}
+    ~TestICD() {}
+
+    TestICD& SetMinICDInterfaceVersion(uint32_t icd_interface_version) {
+        this->min_icd_interface_version = icd_interface_version;
+        return *this;
+    }
+    TestICD& SetICDAPIVersion(uint32_t api_version) {
+        this->icd_api_version = api_version;
+        return *this;
+    }
+    TestICD& AddInstanceLayer(Layer layer) { return AddInstanceLayers({layer}); }
+
+    TestICD& AddInstanceLayers(std::vector<Layer> layers) {
+        this->instance_layers.reserve(layers.size() + this->instance_layers.size());
+        for (auto& layer : layers) {
+            this->instance_layers.push_back(layer);
+        }
+        return *this;
+    }
+    TestICD& AddInstanceExtension(Extension extension) { return AddInstanceExtensions({extension}); }
+
+    TestICD& AddInstanceExtensions(std::vector<Extension> extensions) {
+        this->instance_extensions.reserve(extensions.size() + this->instance_extensions.size());
+        for (auto& extension : extensions) {
+            this->instance_extensions.push_back(extension);
+        }
+        return *this;
+    }
+
+    PhysicalDevice& FindPhysDevice(std::string deviceName) {
```

Comment:
This was from when I imagined it being useful to refer to devices by name. I have since dropped that as indexes are simpler for implementations and ad-hoc string lookup is just bad in general.